File: interviews/leetcode/160. Intersection of Two Linked Lists/length_alignment.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

def get_length(head):
    length = 0
    while head:
        length += 1
        head = head.next
    return length


class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        # Calculate lengths of both lists
        lenA, lenB = get_length(headA), get_length(headB)

        # Align starting points
        while lenA > lenB:
            headA = headA.next
            lenA -= 1
        while lenB > lenA:
            headB = headB.next
            lenB -= 1

        # Move both pointers until they meet
        while headA and headB:
            if headA == headB:  # Check if nodes are the same by reference
                return headA
            headA = headA.next
            headB = headB.next

        return None  # No intersection

File: interviews/leetcode/arrays-and-strings/160. Intersection of Two Linked Lists/length_alignment.go
package main

// Definition for singly-linked list.
type ListNode struct {
	Val  int
	Next *ListNode
}

func getListLen(head *ListNode) int {
	length := 0
	for head != nil {
		length++
		head = head.Next
	}
	return length
}

// Function to find the intersection node
func getIntersectionNode(headA, headB *ListNode) *ListNode {

	// Calculate lengths of both lists
	lenA, lenB := getListLen(headA), getListLen(headB)

	// Align the starting points
	for lenA > lenB {
		headA = headA.Next
		lenA--
	}
	for lenB > lenA {
		headB = headB.Next
		lenB--
	}

	// Move both pointers until they meet
	for headA != nil && headB != nil {
		if headA == headB { // Check if nodes are the same by reference
			return headA
		}
		headA = headA.Next
		headB = headB.Next
	}

	return nil // No intersection
}
